### PR TITLE
Optuna/ LossもOptuna

### PR DESCRIPTION
**Pull request by yumion**
### 変更点
- LossもOptunaで探索できるようにしました
- 各Trialでの結果をそれぞれ保存するようにしました
フォルダ構成は以下
     result_dir/trial#/image/, learning/, model/
     result_dir/info/

- GPUメモリを節約
使う分だけメモリを確保するコードを `set_train_env_task.py` に追記しました

- train中にCtrl+cで学習を止める時，last_model.h5として保存するようにしました
fitにtry/exceptを噛ませました
ただし，trainデータをtrain中のみに働きます

### 問題
- optunaのlogの保存場所が指定できません
train_workflow.pyの
```
logger.addHandler(logging.FileHandler(
            "optuna_trials.log", mode="w")
        )
```
でlogの保存場所を指定しますが，なぜか相対パス絶対パス共に指定してもそこにファイルが保存されておらず，ファイル名だけ書いた場合保存されることを確認しています．
解決求．

- Classificationは対応していません（一部対応していますが動かしていないので保証できません）

### CODE REVIEW

**Comment by yumion** [comment]
Optunaが正しく動いてないっぽいです
loss/lr変わってもtestのDiceの値が変わらない
（backboneが変わるとdiceも変わる）

File: farmer/api/main.py [hunk 1]
```diff
@@ -68,14 +72,82 @@ def fit():
                 trainer.learning_path = f"{k_result}/{trainer.learning_dir}"
                 trainer.image_path = f"{k_result}/{trainer.image_dir}"
 
-                train_workflow = TrainWorkflow(trainer)
                 if trainer.optuna:
-                    train_workflow.optuna_command()
```
Comment (yumion) [inline farmer/api/main.py:73]: optunaのlogを保存する
```
        # Logger for Optuna
        import logging
        logger = logging.getLogger()
        logger.setLevel(logging.INFO)  # Setup the root logger.
        # TODO: 任意のパスを設定したら保存されない問題．infoフォルダにいれたい．
        logger.addHandler(logging.FileHandler(
            "optuna_trials.log", mode="w")
        )
        optuna.logging.enable_propagation()  # Propagate logs to the root logger.
        optuna.logging.enable_default_handler()  # Stop showing logs in sys.stderr.
```

File: farmer/domain/workflows/train_workflow.py [hunk 2]
```diff
@@ -10,54 +12,100 @@
...
+        
+            # set train params to params setted by optuna
+            self._config.train_params = set_train_params(self._config.optuna_params)
+        
     def command(self, trial=None):
         self.set_env_flow()
         train_set, validation_set, test_set = self.read_annotation_flow()
-        self.eda_flow()
-        model, base_model = self.build_model_flow(trial)
+        self.eda_flow(train_set)
```
Comment (yumion) [inline farmer/domain/workflows/train_workflow.py:67]: 初めのtrialの時だけEDAする
```
if trial is None or trial.number == 0:
    self.eda_flow(train_set)
```

File: farmer/domain/workflows/train_workflow.py [hunk 3]
```diff
@@ -10,54 +12,100 @@
...
+            # result_dir/trial#/image/
+            self._config.image_path = self._config.image_path.replace("/image", f"/trial{trial.number}/image")
+
+            self._config.trial_number = trial.number
+            self._config.trial_params = trial.params
+
+            def set_train_params(params_dict: dict) -> dict:
+                params = {}
+                for key, val in params_dict.items():
+                    if not type(val) in {list, dict}:
```
Comment (yumion) [inline farmer/domain/workflows/train_workflow.py:38]: isinstanceを使うとこう書けるみたいです
```
if isinstance(val, (list, dict)):
```

File: farmer/domain/workflows/train_workflow.py [hunk 4]
```diff
@@ -10,54 +12,100 @@
...
+                        params[key] = val
+                    elif type(val) == list:
+                        if type(val[0]) == str:
```
Comment (yumion) [inline farmer/domain/workflows/train_workflow.py:41]: 多分 `is` が一般的です
```
if type(val[0]) is str:
```

File: farmer/domain/workflows/train_workflow.py [hunk 5]
```diff
@@ -10,54 +12,100 @@
...
+                            params[key] = trial.suggest_categorical(
+                                f'{key}', val
```
Comment (yumion) [inline farmer/domain/workflows/train_workflow.py:43]: `key` は元から stringじゃないんですか？
```
params[key] = trial.suggest_categorical(key, val)
```
じゃダメなんですか？

File: farmer/ncc/losses/loss_functions.py [hunk 6]
```diff
@@ -4,40 +4,47 @@
...
+
 def dice_loss(beta=1, class_weights=None, class_indexes=None, **kwargs):
     loss = DiceLoss(
-        beta=beta, 
-        class_weights=class_weights, 
+        beta=beta,
+        class_weights=class_weights,
         class_indexes=class_indexes
     )
     return loss
```
Comment (yumion) [inline farmer/ncc/losses/loss_functions.py:19]: `loss` に渡さないでそのままインスタンス返してはだめでしょうか？（好みだと思いますが）
